File: backend/app/com/skewwhiffy/auraltester/scales/Key.scala
package com.skewwhiffy.auraltester.scales

import com.skewwhiffy.auraltester.notes.{AbsoluteNote, Accidental, Note}
import com.skewwhiffy.auraltester.scales.Key.renderableKeys

object Key {
  private lazy val renderableKeys: List[String] = "C G D A E B F# C# F Bb Eb Ab Db Gb Cb".split(' ').toList
  lazy val cMajor: Key = new Key(Note.c)
}

class Key(val note: Note, val isMinor: Boolean = false) {
  lazy val abc: String = if (isMinor) s"${note.displayString}m" else note.displayString

  def abc(note: AbsoluteNote): String = {
    s"${accidentalAbc(note.note)}${noteAbc(note)}"
  }

  def canRenderSignature: Boolean = {
    if (isMinor) {
      return relativeMajor.canRenderSignature
    }
    renderableKeys.contains(note.displayString)
  }

  private def noteAbc(note: AbsoluteNote): String = {
    note.octave.offsetFromDefault match {
      case 0 => note.note.noteName
      case it if it > 0 => s"${note.note.noteName.toLowerCase}${"'".repeat(it - 1)}"
      case it if it < 0 => s"${note.note.noteName}${",".repeat(-it)}"
    }
  }

  private def accidentalAbc(note: Note): String = {
    notes.find(it => it.noteName == note.noteName) match {
      case None => throw new IllegalArgumentException(s"No note ${note.noteName} in scale")
      case Some(it) => note.accidental match {
        case it.accidental => ""
        case Accidental.natural => "="
        case it => it.abc
      }
    }
  }

  def relativeMinor: Key = if (isMinor) this else new Key(note.downMajorSecond.downMajorSecond.sharp, true)

  def relativeMajor: Key = if (isMinor) new Key(note.upMajorSecond.upMajorSecond.flat, false) else this

  def notes: List[Note] = {
    (if (isMinor) List(2, 1, 2, 2, 1, 2) else List(2, 2, 1, 2, 2, 2))
      .scanLeft(note) {
        case (it, 2) => it.upMajorSecond
        case (it, 1) => it.upMajorSecond.flat
      }
  }

  override def equals(obj: Any): Boolean = obj match {
    case other: Key => note == other.note && isMinor == other.isMinor
    case _ => false
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/key/Key.kt
package com.skewwhiffy.notation.model.key

import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.notation.model.note.Accidental
import com.skewwhiffy.notation.model.note.Note
import java.lang.IllegalArgumentException

data class Key(val note: Note, val isMinor: Boolean = false) {
  companion object {
    @Suppress("SpellCheckingInspection")
    private val renderableKeys: List<String> = "C G D A E B F# C# F Bb Eb Ab Db Gb Cb"
      .split(' ')
      .toList()

    val cMajor: Key = Key(Note.c)
  }

  val abc: String = if (isMinor) "${note.displayString}m" else note.displayString

  val displayString: String = "${note.displayString} ${if (isMinor) "minor" else "major"}"

  fun abc(note: AbsoluteNote): String {
    return "${accidentalAbc(note.note)}${noteAbc(note)}"
  }

  val canRenderSignature: Boolean = if (isMinor) relativeMajor.canRenderSignature
  else renderableKeys.contains(note.displayString)

  private fun noteAbc(note: AbsoluteNote): String {
    return note.octave.offsetFromDefault.let {
      when {
        it > 0 -> "${note.note.noteName.lowercase()}${"'".repeat(it - 1)}"
        it < 0 -> "${note.note.noteName}${",".repeat(-it)}"
        else -> note.note.noteName
      }
    }
  }

  private fun accidentalAbc(note: Note): String {
    return notes.find { it.noteName == note.noteName }.let {
      it ?: throw IllegalArgumentException("No note ${note.noteName} in scale")
      when (note.accidental) {
        it.accidental -> ""
        Accidental.natural -> "="
        else -> note.accidental.abc
      }
    }
  }

  val relativeMinor: Key
    get() = if (isMinor) this else Key(
      note.downMajorSecond.downMajorSecond.sharp,
      isMinor = true
    )

  val relativeMajor: Key get() = if (isMinor) Key(note.upMajorSecond.upMajorSecond.flat) else this

  val relative: Key get() = if (isMinor) relativeMajor else relativeMinor

  val notes: List<Note>
    get() {
      return (if (isMinor) listOf(2, 1, 2, 2, 1, 2) else listOf(2, 2, 1, 2, 2, 2))
        .scan(note) { note: Note, semitones: Int ->
          if (semitones == 2) note.upMajorSecond else note.upMajorSecond.flat
        }
    }

  /*
    override def equals(obj: Any): Boolean = obj match {
      case other: Key => note == other.note && isMinor == other.isMinor
      case _ => false
    }
  }
   */
}
